2017-00-00-shipping-rust-1.md

```diff
@@ -0,0 +1,70 @@
+# Shipping Rust on Tessel
+
+***Tessel is an open-source hardware platform for developing IoT devices.***
+
+The Tessel board runs an embedded Linux distribution (OpenWRT) and uses a less-common chip architecture: MIPS with soft-float.
+Our CLI is designed to require as few steps as possible between plugging in the Tessel board and running code. The board was designed to run Node.js, a runtime for JavaScript, which is an interpreted language. Beyond pruning the amount of code the user has downloaded to that which is actually needed on the device, essentially we just need to tarball a user’s code, unpack it, and run it behind the scenes over an SSH-like pipe:
+
+```
+$ t2 init
+$ t2 run index.js
+Blinking!
+```
+
+When we wanted to expand the CLI tool’s support for the Rust programming language, we had to rethink interactions with the command line tool—instead of running a script directory, you specify the target binary by name, for example—but none were more involved than cross-compilation. Tessel’s architecture is not one compiled by default for Rust, so we have to ship the pieces we need ourselves.
+
+The cross-compilation process for Rust involves the following:
+
+- Cross-compile the user’s Rust code. This is specified by the LLVM configuration in a target.json file, or the built-in specifications for several architectures (e.g. x86–64-darwin)
+- Search for a libstd for the target platform. This libstd is compiled and provided for the user. You can pull down the target libraries for a set of platforms using `rustup target add ...`.
+- Link these together using a linker for the target platform, e.g. `mipsel-openwrt-linux-ld`.
+
+So this is two binaries we need to ship, as well as one JSON file.
+
+First is libstd. We actually want to compile the Rust standard libraries against our target.json, but here we have a problem: how do you do this? … … We use [rust-cross-libs.sh](https://medium.com/r/?url=https%3A%2F%2Fgithub.com%2Ftessel%2Ft2-rustlib%2Fblob%2Fmaster%2Frust-cross-libs.sh), a script to (with some effort) cross-compile the libstd and generate a new libstd according to the target.json file.
```

Comment:
The makefile-based build system that rust-cross-libs.sh uses is [about to be removed](https://internals.rust-lang.org/t/proposal-for-promoting-rustbuild-to-official-status/4368). I doubt that script works unmodified to build Rust 1.14 anyway, though...